### ISSUE TITLE
Query Parallelization Tracking Issue

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#48685]
This issue is a sub-issue of https://github.com/rust-lang/rust/issues/48547: it tracks the in-progress effort to parallelize rustc across queries. This work is being spearheaded by @zoxc. 

### Goals

Allow rustc to execute queries in parallel with one another. Enable the use of rayon or other tools for intra-query parallelization as well. See [this internals thread](https://internals.rust-lang.org/t/parallelizing-rustc-using-rayon/6606) for more information.

### Overview of the plan

- [ ] Make all types in the compiler send and sync by landing @Zoxc's existing work
    - The existing work largely replaces `RefCell` usage with `Mutex`
    - In many cases, [the interior mutability itself can be refactored away](https://internals.rust-lang.org/t/parallelizing-rustc-using-rayon/6606/24?u=nikomatsakis). The plan is do this as a second step.
- [ ] *write me*

### Refactoring list

This section tracks refactorings that we should do to remove mutex usage.

**Comment by nikomatsakis** [comment]
cc @michaelwoerister

**Comment by Zoxc** [comment]
@michaelwoerister We also have to make queries work on a Rayon thread-pool

**Comment by michaelwoerister** [comment]
> We also have to make queries work on a Rayon thread-pool

What exactly does this involve?

**Comment by Zoxc** [comment]
@michaelwoerister We have to make waiting on already executing queries block and also detect and recover from query cycles

**Comment by michaelwoerister** [comment]
Ah, right. The blocking should be easy enough. I have not thought in detail about cycle detection and recovery though.

**Comment by michaelwoerister** [comment]
> We have to make waiting on already executing queries block and also detect and recover from query cycles

Both of these should be implemented by #50699, right?

**Comment by Zoxc** [comment]
@michaelwoerister Yeah.

**Comment by SparrowLii** [comment]
@nikomatsakis Some refactoring work has been done. How should we update the task list?

**Comment by petrochenkov** [comment]
Closing this in favor of https://github.com/rust-lang/rust/issues/113349, that issue tracks more recent development in this area.
